src/extras/render-passes/render-pass-depth-aware-blur.js

```diff
@@ -1,93 +1,197 @@
+        tap(sum, totalWeight, kernel[middle - i], depth, uv0 - float(i) * direction * sourceInvResolution);
+    }
+
+    float finalColor = sum / totalWeight;
+
+    // simple dithering helps a lot (assumes 8 bits target)
+    // this is most useful with high quality/large blurs
+    // finalColor += ((random(gl_FragCoord.xy) - 0.5) / 255.0);
+
+    gl_FragColor.r = finalColor;
+}
+`;
+
 /**
  * Render pass implementation of a depth-aware bilateral blur filter.
  *
  * @category Graphics
  * @ignore
  */
 class RenderPassDepthAwareBlur extends RenderPassShaderQuad {
-    constructor(device, sourceTexture) {
-        super(device);
-        this.sourceTexture = sourceTexture;
+    _direction = new Float32Array(2);
+
+    /**
+     * Initializes a new blur render pass. It has to be called before the render pass can be used.
+     *
+     * @param {Object} options - Options for configuring the blur effect.
+     * @param {import('../../platform/graphics/texture.js').Texture} [options.sourceTexture] - The source texture to be blurred.
+     * @param {number} [options.kernelSize] - The size of the blur kernel. Defaults to 7.
+     * @param {BLUR_GAUSSIAN|import('../../scene/constants.js').BLUR_BOX} [options.type] - The type of blur to apply. Defaults to BLUR_GAUSSIAN.
+     * @param {Vec2} [options.direction] - The direction of the blur. Defaults to (1, 0).
+     */
+    setup(options) {
+        this.sourceTexture = options.sourceTexture;
+
+        this.direction = options.direction ?? new Vec2(1, 0);
+        this.kernelSize = options.kernelSize ?? 9;
+        this.type = options.type ?? BLUR_GAUSSIAN;
 
-        this.shader = this.createQuadShader('DepthAwareBlurShader', shaderChunks.screenDepthPS + /* glsl */`
+        var scope = this.device.scope;
 
-            varying vec2 uv0;
+        this.sourceTextureId = scope.resolve('sourceTexture');
+        this.sourceInvResolutionId = scope.resolve('sourceInvResolution');
+        this.kernelId = scope.resolve('kernel[0]');
+        this.directionId = scope.resolve('direction');
+    }
 
-            uniform sampler2D sourceTexture;
-            uniform vec2 sourceInvResolution;
-            uniform int filterSize;
+    execute() {
+        if (!this.sourceTexture) {
+            super.execute();
+            return;
+        }
 
-            float random(const highp vec2 w) {
-                const vec3 m = vec3(0.06711056, 0.00583715, 52.9829189);
-                return fract(m.z * fract(dot(w, m.xy)));
-            }
+        const { width, height } = this.sourceTexture;
+        sourceInvResolutionValueTmp[0] = 1.0 / width;
+        sourceInvResolutionValueTmp[1] = 1.0 / height;
 
-            float bilateralWeight(in float depth, in float sampleDepth) {
-                float diff = (sampleDepth - depth);
-                return max(0.0, 1.0 - diff * diff);
-            }
+        this.sourceInvResolutionId?.setValue(sourceInvResolutionValueTmp);
```

Comment:
no need for those `?` there as those are transpiled to `if` .. we always set those up, so they're never undefined.

Reply:
VS Code TS checker highlighted this error, suggesting that sourceInvResolutionId could be undefined as it is not defined during creation. I'll check if Playcanvas linter highlights it.